chain/pruned_block_dispatcher.go

```diff
@@ -0,0 +1,528 @@
+}
+
+// PrunedBlockDispatcher enables a chain client to request blocks that the
+// server has already pruned. This is done by connecting to the server's full
+// node peers and querying them directly. Ideally, this is a capability
+// supported by the server, though this is not yet possible with bitcoind.
+type PrunedBlockDispatcher struct {
+	cfg PrunedBlockDispatcherConfig
+
+	// workManager handles satisfying all of our incoming pruned block
+	// requests.
+	workManager *query.WorkManager
+
+	// blocksQueried represents the set of pruned blocks we've been
+	// requested to query. Each block maps to a list of clients waiting to
+	// be notified once the block is received.
+	//
+	// NOTE: The blockMtx lock must always be held when accessing this
+	// field.
+	blocksQueried map[chainhash.Hash][]chan *wire.MsgBlock
+	blockMtx      sync.Mutex
+
+	// currentPeers represents the set of peers we're currently connected
+	// to. Each peer found here will have a worker spawned within the
+	// workManager to handle our queries.
+	//
+	// NOTE: The peerMtx lock must always be held when accessing this
+	// field.
+	currentPeers map[string]struct{}
+	peerMtx      sync.Mutex
+
+	// peersConnected is the channel through which we'll send new peers
+	// we've established connections to.
+	peersConnected chan query.Peer
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewPrunedBlockDispatcher initializes a new PrunedBlockDispatcher instance
+// backed by the given config.
+func NewPrunedBlockDispatcher(cfg *PrunedBlockDispatcherConfig) (
+	*PrunedBlockDispatcher, error) {
+
+	if cfg.NumTargetPeers < 1 {
+		return nil, errors.New("config option NumTargetPeer must be >= 1")
+	}
+
+	peersConnected := make(chan query.Peer)
+	return &PrunedBlockDispatcher{
+		cfg: *cfg,
+		workManager: query.New(&query.Config{
+			ConnectedPeers: func() (<-chan query.Peer, func(), error) {
+				return peersConnected, func() {}, nil
+			},
+			NewWorker: query.NewWorker,
+			Ranking:   query.NewPeerRanking(),
+		}),
+		blocksQueried:  make(map[chainhash.Hash][]chan *wire.MsgBlock),
+		currentPeers:   make(map[string]struct{}),
+		peersConnected: peersConnected,
+		quit:           make(chan struct{}),
+	}, nil
+}
+
+// Start allows the PrunedBlockDispatcher to begin handling incoming block
+// requests.
+func (d *PrunedBlockDispatcher) Start() error {
+	log.Tracef("Starting pruned block dispatcher")
+
+	if err := d.workManager.Start(); err != nil {
+		return err
+	}
+
+	d.wg.Add(1)
+	go d.pollPeers()
+
+	return nil
+}
+
+// Stop stops the PrunedBlockDispatcher from accepting any more incoming block
+// requests.
+func (d *PrunedBlockDispatcher) Stop() {
+	log.Tracef("Stopping pruned block dispatcher")
+
+	close(d.quit)
+	d.wg.Wait()
+
+	_ = d.workManager.Stop()
+}
+
+// pollPeers continuously polls the backend node for new peers to establish
+// connections to.
+func (d *PrunedBlockDispatcher) pollPeers() {
+	defer d.wg.Done()
+
+	if err := d.connectToPeers(); err != nil {
```

Comment:
Do we need to do this if there aren't actually any blocks that need fetching yet? Otherwise we'll create some TCP connections (and keep them alive?) that we'll never actually need to use (say it's a full full node).